code/game/machinery/mind_machine.dm

```diff
@@ -0,0 +1,846 @@
+		else
+			mindTypeOne = "Higher" //Player controlled
+	if(isvampire(S) || isanycultist(S) || ischangeling(S) || ismalf(S))
+		mindTypeOne = "Shielded" //Mostly to fix spell bugs but also tinfoil
+	if((ishigherbeing(S)) || (ismonkey(S)))
+		var/mob/living/carbon/T = S
+		if(T.is_wearing_item(/obj/item/clothing/head/tinfoil))
+			mindTypeTwo = "Shielded"
+	S = occupantTwo
+	occupantNameTwo = S.name //and for pod two
+	switch(S.stat)
+		if(0)
+			occupantStatTwo = "Alive"
+		if(1)
+			occupantStatTwo = "Unconscious"
+		if(2)
+			occupantStatTwo = "Dead"
+	if(!S.mind)
+		mindTypeTwo = "Lower"
+	if(S.mind)
+		if(isrobot(S))
+			mindTypeTwo = "Artificial"
+		else
+			mindTypeTwo = "Higher"
+	if(isvampire(S) || isanycultist(S) || ischangeling(S) || ismalf(S))
+		mindTypeTwo = "Shielded"
+	if((ishigherbeing(S)) || (ismonkey(S)))
+		var/mob/living/carbon/T = S
+		if(T.is_wearing_item(/obj/item/clothing/head/tinfoil))
+			mindTypeTwo = "Shielded"
+	currentlySwapping = FALSE
+
+/obj/machinery/mind_machine/mind_machine_hub/proc/swapOccupants(var/mob/living/M)
+	if(occupantScan != TRUE || lockedPods != TRUE || badSwap == TRUE)
+		return
+	if((!locate(occupantOne,connectOne.contents)) || (!locate(occupantTwo, connectTwo.contents)))
+		errorMessage = "Occupants missing" //sanity and wizards exist
+		occupantOne = null
+		occupantTwo = null
+		unlockPods()
+		return
+	if(bluespaceConduit == 0)
+		errorMessage = "Conduit requires charge"
+		return
+	if(mindTypeOne == "Shielded" || mindTypeTwo == "Shielded")
+		errorMessage = "Critical error, aborting"
+		spark(src)
+		spark(connectOne)
+		spark(connectTwo)
+		unlockPods()
+		return
+	if(occupantStatOne == "Dead" || occupantStatTwo == "Dead")
+		if(!soulShardSafety) //Secrets
+			errorMessage = "Living mind required"
+			return
+		soulShardSafety = FALSE
+		errorMessage = "Warning: Unknown internal reaction"
+	currentlySwapping = TRUE
+	bluespaceConduit -= 1
+	icon_state = "mind_hub_active"
+	connectOne.icon_state = "mind_pod_active"
+	connectTwo.icon_state = "mind_pod_active"
+	if(is_type_in_list(occupantOne, illegalSwap) || is_type_in_list(occupantTwo, illegalSwap))
+		malfSwap = TRUE //fun sanity check. Can't send their mind into the illegal body.
+	M = occupantOne //First we check if they have any pocket-pets
+	for(var/obj/item/weapon/holder/i in get_contents_in_object(M))  //Carp in pack? Outta luck, Jack
+		theFly += i
+	var/list/borers = M.get_brain_worms()
+	for(var/mob/living/simple_animal/borer/B in borers) //borers in the body are counted
+		theFly += B
+	M = occupantTwo
+	for(var/obj/item/weapon/holder/i in get_contents_in_object(M))
+		theFly += i
+	borers += M.get_brain_worms()
+	for(var/mob/living/simple_animal/borer/B in borers)
+		theFly += B
+	for(var/prog in 1 to 40/manipRating) //Counts up 5 to 10 seconds, checks if we lost power each time and ruins your day if so. Otherwise just for UI progress bar
+		if(stat & NOPOWER)
+			malfSwap = TRUE
+		sleep(1 SECONDS)
+		swapProgress += 1
```

Comment:
You're not calling `nanomanager.update_uis(src)` here so the UI won't update correctly.

Reply:
I have it set to update automatically as long as it's open. In regard to UI updates that all definitely works.

Reply:
The problem with that is it still can cause a maximum of 2 seconds of latency before the UI updates, because that's how slow NanoUI's auto update is.

In this case I guess that's fine, but in a bunch of the others it isn't and can cause noticeable, painful UI lag.